server2/Arcadia.Assistant/Arcadia.Assistant.ExternalStorages.SharepointOnline/Contracts/SharepointRequest.cs

```diff
@@ -0,0 +1,139 @@
+﻿namespace Arcadia.Assistant.ExternalStorages.SharepointOnline.Contracts
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Net.Http;
+    using System.Net.Http.Headers;
+    using System.Text.Json;
+
+    public class SharepointRequest
+    {
+        private const string AcceptHeaderName = "Accept";
+        private const string AuthorizationHeaderName = "Authorization";
+        private const string ContentTypeHeaderName = "Content-Type";
+        private const string ContentLengthHeaderName = "Content-Length";
+        private const string IfMatchHeaderName = "IF-MATCH";
+        private const string XHttpMethodHeaderName = "X-HTTP-Method";
+
+        private readonly List<Tuple<string, string>> headersInternal = new List<Tuple<string, string>>();
+
+        private SharepointRequest(HttpMethod httpMethod, string url)
+        {
+            this.HttpMethod = httpMethod;
+            this.Url = url;
+        }
+
+        public HttpMethod HttpMethod { get; }
+
+        public string Url { get; }
+
+        public HttpContent? Content { get; private set; }
+
+        public IReadOnlyList<Tuple<string, string>> Headers => this.headersInternal;
+
+        public static SharepointRequest Create(HttpMethod httpMethod, string url)
+        {
+            return new SharepointRequest(httpMethod, url);
+        }
+
+        public SharepointRequest WithAcceptHeader(string value, bool update = true)
```

Comment:
Do you use it anywhere with `update` argument set to `false`?

Reply:
Previous implementation work as update = false. Currently it used as update = true, but you can add several Accept header with different values. Current signature is correct.

Reply:
Current signature is excessive, because you don't develop a public library where you have to forecast all possible cases, but an internal class which is used only in other internal classes. And you know all the use cases. And this use case is not needed in current version of code, so this code is "just in case" which is not needed.

Reply:
I agree with you partially. But you incapsulate Headers construction logic to separate class. You have no possibility to manage ones outside. Some cases require to update this header, another - to add the same header. You have to make both request header types event todays build doesnt use second.

Reply:
The key words - "todays build doesn't use second". Only you decide when you use or don't use some use cases. Now you don't use it - then it is not needed at all and could be removed. When you need it - it can be added either from the scratch or from repository history.

Reply:
Disagree. It's ok for inner functions and classes. We implement some balck box with interface. I cant lose my time for looking up incorrect black box behaviour and fix it for my new functionality (but i did it with this code). Your description is good for single programmist code writing. It can affort group developing time. Ok, i can remoe this option, but i publish headers collection too.

Reply:
Black box is a library which you don't have access to. Here you can open code of `WithAcceptHeader` method and find out that it adds a new header, not replaces an existing one.
By the way, why didn't you add `update` param to each `With...Header` method? Maybe, because it is not needed? Here it is not needed too, you can always remove old header and add a new one, because you know that accept header could be only single. And authorization header could be only one, and all others used in this code.

Reply:
Incapsulation - the main way to make 'Black box' - you exclude some finally writed code to external object for use it in any places. Even it some 'Generator' in you own lib - the main idea - to create some engine (box) for use throught external interface. If you rewrite it every time - something wrong in your app architecture.
Ok, i heard you. According rfc7230 this SharepointRequest class write tottaly incorrect. I will rewrite it. I will lose my time for already writed code - it's worst code practise with good theoretic base.

Reply:
> the main way to make 'Black box'

Yes, sure. But the main difference with external library which you are not responsible for - you are able to make changes to that "Black box" at any time. When you need such functionality - you can add it. If you don't need it - it doesn't make sense to have it just "because one day in the future we, probably, will use it or maybe won't".

> If you rewrite it every time - something wrong in your app architecture.

Not any time, but only when you need to add some functionality that wasn't implemented yet. Or do you think that external libraries are written only once without rewriting?

> According rfc7230 this SharepointRequest class write tottaly incorrect

Which exact point of that RFC is violated?

> I will rewrite it

I don't ask you to do this, I ask just to remove not used parameter. That is a good point that headers should not multiply, but they should not multiply without relation to the value of the param, but always.

Reply:
Code rewrited.